### PR TITLE
Update Segment calls

### PR DESCRIPTION
**Pull request by TimJones97**
## Motivation
A Cookie Consent banner must be shown to allow the user to select mandatory cookies only (`Reject` action) or all tracking cookies (`Accept` action). The Identify and Login calls also require updating to allow tracking more granular data points.

https://mycelium-mex.atlassian.net/browse/SWAP-148 - Swaps Cookie Notice and Consent Management
https://mycelium-mex.atlassian.net/browse/SWAP-150 - Swaps Identify Call

## Changelog

- Add Cookie Consent banner
- Use localStorage item to store `consentAcknowledged` 
- Update `segmentAnalytics.js` component to create new context provider for analytics function access across app
- Add `trackLogin()` function to send `walletAddress`, `walletProvider`, `network`, `ethBalance` and `gmxBalances` to Segment which triggers on wallet connection (if consent approved)

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -353,9 +364,26 @@ function AppHeaderUser({
...
+  const tokens = getTokens(chainId);
+  const tokenAddresses = tokens.map((token) => token.address);
+
+  const { data: tokenBalances } = useSWR(
+    [`GlpSwap:getTokenBalances:${active}`, chainId, readerAddress, "getTokenBalances", account || PLACEHOLDER_ACCOUNT],
+    {
+      fetcher: fetcher(library, ReaderV2, [tokenAddresses]),
+    }
+  );
+  const { infoTokens } = useInfoTokens(library, chainId, active, tokenBalances, undefined);
```
Comment (dospore) [inline src/App.js:383]: This is a round about way to get the eth balance. You can use `library.getBalance(account)` from web3React

File: src/segmentAnalytics.js [hunk 2]
```diff
@@ -1,20 +1,105 @@
...
+import { useLocation } from "react-router-dom";
+import { ARBITRUM, ARBITRUM_TESTNET, AVALANCHE, CURRENT_PROVIDER_LOCALSTORAGE_KEY, hasUserConsented } from "./Helpers";
+import { useWeb3React } from "@web3-react/core";
+
+const writeKey = process.env.REACT_APP_SEGMENT_WRITE_KEY;
+
+const networkName = {
+  [ARBITRUM]: "Arbitrum",
+  [ARBITRUM_TESTNET]: "Rinkeby",
+  [AVALANCHE]: "Avalanche",
```
Comment (dospore) [inline src/segmentAnalytics.js:12]: Can remove avalanche here

File: src/segmentAnalytics.js [hunk 3]
```diff
@@ -1,20 +1,105 @@
...
+  const [analytics, setAnalytics] = useState(undefined);
+
+  const trackLogin = (chainId, gmxBalances, balanceEth) => {
+    const hasConsented = hasUserConsented();
+    try {
+      const provider = localStorage.getItem(CURRENT_PROVIDER_LOCALSTORAGE_KEY);
+      const traits = {
+        walletProvider: provider,
+        walletAddress: account,
+        network: networkName[chainId],
```
Comment (dospore) [inline src/segmentAnalytics.js:33]: Not necessary but will be undefined if networkName does not have an entry for that chainId 
```suggestion
        network: networkName[chainId] ?? `Unsupported (${chainId})`,
```

File: public/index.html [hunk 4]
```diff
@@ -5,43 +5,34 @@
...
+    <meta name="msapplication-TileImage" content="/favicon/ms-icon-144x144.png" />
+    <meta name="theme-color" content="#ffffff" />
     <title>Tracer Perpetual Swaps</title>
-    <script type="text/partytown">
-      !function(){var analytics=window.analytics=window.analytics||[];if(!analytics.initialize)if(analytics.invoked)window.console&&console.error&&console.error("Segment snippet included twice.");else{analytics.invoked=!0;analytics.methods=["trackSubmit","trackClick","trackLink","trackForm","pageview","identify","reset","group","track","ready","alias","debug","page","once","off","on"];analytics.factory=function(t){return function(){var e=Array.prototype.slice.call(arguments);e.unshift(t);analytics.push(e);return analytics}};for(var t=0;t<analytics.methods.length;t++){var e=analytics.methods[t];analytics[e]=analytics.factory(e)}analytics.load=function(t,e){var n=document.createElement("script");n.type="text/partytown";n.async=!0;n.src="https://cdn.segment.com/analytics.js/v1/"+t+"/analytics.min.js";var a=document.getElementsByTagName("script")[0];a.parentNode.insertBefore(n,a);analytics._loadOptions=e};analytics.SNIPPET_VERSION="4.1.0";
```
[the remark this answers is not in the document]
Comment (TimJones97) [reply public/index.html:31]: Hey, I saw that, and I tried to use it. Buy unfortunately couldn't get analytics.track to work without the new method

[the remark this answers is not in the document]
Comment (TimJones97) [reply public/index.html:31]: If we remove the last useEffect and keep the snippet wouldn't it be loading twice then?

File: src/App.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import React, { useState, useEffect, useCallback, useRef } from "react";
+import React, { useState, useEffect, useCallback, useRef, useContext } from "react";
```
Comment (dospore) [inline src/App.js:1]: Unused?
```suggestion
import React, { useState, useEffect, useCallback, useRef } from "react";
```

File: src/App.js [hunk 6]
```diff
@@ -353,9 +362,23 @@ function AppHeaderUser({
...
+
   const exchangeRef = useRef();
   const { connector, library, deactivate, activate, active, account } = useWeb3React();
   const { chainId } = useChainId();
+  const readerAddress = getContract(chainId, "Reader");
+  const tokens = getTokens(chainId);
+  const tokenAddresses = tokens.map((token) => token.address);
+
+  const { data: tokenBalances } = useSWR(
+    [`GlpSwap:getTokenBalances:${active}`, chainId, readerAddress, "getTokenBalances", account || PLACEHOLDER_ACCOUNT],
```
Comment (dospore) [inline src/App.js:376]:
```suggestion
    [`FullApp:getTokenBalances:${active}`, chainId, readerAddress, "getTokenBalances", account || PLACEHOLDER_ACCOUNT],
```

File: src/App.js [hunk 7]
```diff
@@ -367,17 +390,28 @@ function FullApp() {
...
+          if (balanceData[token]) {
+            const fieldName = `balance${formatTitleCase(token)}`;
+            gmxBalances[fieldName] = parseFloat(formatAmount(balanceData[token], MAX_DECIMALS, 4, true));
+          }
+        });
+        // Get user ETH balances
+        const balanceEth = await library.getBalance(account);
+        const formattedEthBalance = parseFloat(formatAmount(balanceEth, MAX_DECIMALS, 4, true)) / 100;
+        trackLogin(chainId, gmxBalances, formattedEthBalance);
+        setLoggedInTracked(true); // Only track once
```
Comment (dospore) [inline src/App.js:410]: do we want this to re-track on account change?

File: src/App.js [hunk 8]
```diff
@@ -1021,9 +1062,12 @@ function App() {
       <Web3ReactProvider getLibrary={getLibrary}>
         <SEO>
           <ThemeProvider>
-            <FullApp />
+            <AnalyticsProvider>
+              <FullApp />
+            </AnalyticsProvider>
```
Comment (dospore) [inline src/App.js:1067]: Can remove this now

File: src/App.js [hunk 9]
```diff
@@ -10,6 +10,8 @@ import { Web3Provider } from "@ethersproject/providers";
 import { Switch, Route, NavLink } from "react-router-dom";
 
 import { ThemeProvider } from "@tracer-protocol/tracer-ui";
+import { useAnalytics, AnalyticsProvider } from "./segmentAnalytics";
```
Comment (dospore) [inline src/App.js:13]: Can remove the provider

**Review by dospore — APPROVED** [review]
Code lgtm will create a seperate pr to remove partytown

**Pull request merged**